cyfra-spirv-tools/src/main/scala/io/computenode/cyfra/spirvtools/SpirvTool.scala

```diff
@@ -0,0 +1,140 @@
+      case Some(file) =>
+        logger.debug(s"Found SPIR-V tool: ${file.getAbsolutePath}")
+        Right(file)
+      case None =>
+        Left(SpirvToolNotFound(toolName))
+    }
+  }
+
+  protected def executeSpirvCmd(shaderCode: ByteBuffer, cmd: Seq[String]): Either[SpirvError, (ByteArrayOutputStream, ByteArrayOutputStream, Int)] = {
+    logger.debug(s"SPIR-V cmd $cmd.")
+    val inputBytes = {
+      val arr = new Array[Byte](shaderCode.remaining())
+      shaderCode.get(arr)
+      shaderCode.rewind()
+      arr
+    }
+    val inputStream = new ByteArrayInputStream(inputBytes)
+    val outputStream = new ByteArrayOutputStream()
+    val errorStream = new ByteArrayOutputStream()
+
+    def safeIOCopy(inStream: InputStream, outStream: OutputStream, description: String): Either[SpirvToolIOError, Unit] = {
+      @tailrec
+      def loopOverBuffer(buf: Array[Byte]): Unit = {
+        val len = inStream.read(buf)
+        if (len == -1) ()
+        else {
+          outStream.write(buf, 0, len)
+          loopOverBuffer(buf)
+        }
+      }
+
+      Using
+        .Manager { use =>
+          val in = use(inStream)
+          val out = use(outStream)
+          val buf = new Array[Byte](1024)
+          loopOverBuffer(buf)
+          out.flush()
+        }
+        .toEither
+        .left
+        .map(e => SpirvToolIOError(s"$description failed: ${e.getMessage}"))
+    }
+
+    def createProcessIO(): Either[SpirvError, ProcessIO] = {
+      val inHandler: OutputStream => Unit =
+        in =>
+          safeIOCopy(inputStream, in, "Writing to stdin") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      val outHandler: InputStream => Unit =
+        out =>
+          safeIOCopy(out, outputStream, "Reading stdout") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      val errHandler: InputStream => Unit =
+        err =>
+          safeIOCopy(err, errorStream, "Reading stderr") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      Try {
+        new ProcessIO(inHandler, outHandler, errHandler)
+      }.toEither.left.map(e => SpirvToolIOError(s"Failed to create ProcessIO: ${e.getMessage}"))
+    }
+
+    for {
+      processIO <- createProcessIO()
+      process <- Try(cmd.run(processIO)).toEither.left.map(ex => SpirvCommandExecutionFailed(s"Failed to execute SPIR-V command: ${ex.getMessage}"))
+    } yield (outputStream, errorStream, process.exitValue())
+  }
+
+  trait SpirvError extends RuntimeException {
+    def message: String
+
+    override def getMessage: String = message
+  }
+
+  protected trait ResultSaveSetting
```

Comment:
Maybe overall it should be called `ToolOutput`? We don't save to logger. If ToLogger is here it can also be `sealed`.